File: medium/438.find-all-anagrams-in-a-string.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

/*
 * @lc app=leetcode id=438 lang=java
 *
 * [438] Find All Anagrams in a String
 */

// @lc code=start
class Solution {
    private int[] targets = new int[26];
    private int[] windows = new int[26];

    public List<Integer> findAnagrams(String s, String p) {
        List<Integer> ans = new ArrayList<Integer>();
        if (s.length() < p.length()) {
            return ans;
        }

        for (char c : p.toCharArray()) {
            targets[c - 'a'] += 1; 
        }
        
        int l = 0;
        int r = 0;
        char[] chars = s.toCharArray();
        
        for (; r<p.length() - 1; r++) {
            windows[chars[r] - 'a'] += 1;
        }

        for (; r<chars.length;l++, r++) {
            windows[chars[r] - 'a'] += 1;
            if (anagram()) {
                ans.add(l);
            }
            windows[chars[l] - 'a'] -= 1;
        }

        return ans;
    }

    private boolean anagram() {
        for (int i=0; i<26; i++) {
            if (targets[i] != windows[i]) {
                return false;
            }
        }
    
        return true;
    }
}
// @lc code=end

File: medium/438.find-all-anagrams-in-a-string.py
#
# @lc app=leetcode id=438 lang=python3
#
# [438] Find All Anagrams in a String
#

# @lc code=start
import collections
from typing import List

class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        def isAnagram():
            for i in range(len(chars_in_p)):
                if chars_in_p[i] != chars_in_window[i]:
                    return False
            return True

        if len(s) < len(p):
            return []

        chars_in_p = [0] * 26
        for c in p:
            chars_in_p[ord(c) - ord('a')] += 1

        chars_in_window = [0] * 26
        r = 0
        while (r < len(p) - 1):
            c = s[r]
            chars_in_window[ord(c) - ord('a')] += 1
            r += 1

        ans = []
        l = 0
        while (r < len(s)):
            chars_in_window[ord(s[r]) - ord('a')] += 1
            if isAnagram():
                ans.append(l)
            chars_in_window[ord(s[l]) - ord('a')] -= 1

            l += 1
            r += 1


        return ans
# @lc code=end

